File: ruby/connect/connect.rb
# With significant gratitude to shybyte
# https://exercism.io/tracks/typescript/exercises/connect/solutions/2e4550709a0241e09842d418c0361225

module DeepCopy
  refine Object do
    def deepcopy
      Marshal.load(Marshal.dump(self))
    end
  end
end

# the Connect game win tester
class Board
  using DeepCopy

  def initialize(input)
    @board = input.map { |row| row.split(' ') }
  end

  def winner
    return 'O' if winner?('O', @board.deepcopy)
    return 'X' if winner?('X', @board.transpose)
    '' # no winner
  end

  private

  def winner?(player, board)
    stack = my_players_in_top_row(player, board)
    return false if stack.empty?
    return true if board.length == 1
    path? player, board, stack
  end

  def my_players_in_top_row(player, board)
    board[0].each_with_index
            .select { |pair| pair.first == player }
            .map { |pair| [0, pair.last] }
  end

  def path?(player, board, stack)
    last_row = board.length - 1
    until stack.empty?
      r, c = stack.pop
      neighbours(player, board, r, c).each do |(rr, cc)|
        return true if rr == last_row

        board[r][c] = 'seen'
        stack << [rr, cc]
      end
    end
    false
  end

  def neighbours(player, board, row, col)
    height = board.length
    width = board[0].length
    neighbours = []

    [-1, 0, 1].each do |dr|
      rr = row + dr
      next if rr.negative? || rr >= height

      [-1, 0, 1].each do |dc|
        cc = col + dc
        next if cc.negative? || cc >= width
        next if dr == dc # cannot move this way

        neighbours << [rr, cc] if board[rr][cc] == player
      end
    end
    neighbours
  end
end

File: javascript/connect/connect.js
/* eslint-disable  no-restricted-syntax, yoda, space-in-parens, no-multi-spaces */
/* eslint no-unused-vars: ["error", { "argsIgnorePattern": "^_$" }] */

const deepCopy = obj => JSON.parse(JSON.stringify(obj));

const transpose = m => m[0].map((_, i) => m.map(row => row[i]));

const neighbours = (player, board, r, c) => {
  const height = board.length;
  const width = board[0].length;
  const result = [];

  for (const dr of [-1, 0, 1]) {
    for (const dc of [-1, 0, 1]) {
      const rr = r + dr;
      const cc = c + dc;
      if (   0 <= rr && rr < height
          && 0 <= cc && cc < width
          && dr !== dc // cannot move this way
          && board[rr][cc] === player
      ) {
        result.push([rr, cc]);
      }
    }
  }
  return result;
};

const isWinner = (player, board) => {
  const b = deepCopy(board);
  const lastRow = b.length - 1;

  // find my players in the top row
  const stack = b[0].reduce((me, v, i) => me.concat(v === player ? [[0, i]] : []), []);

  if (stack.length === 0) return false;
  if (lastRow === 0) return true;

  while (stack.length > 0) {
    const [r, c] = stack.pop();
    for (const [rr, cc] of neighbours(player, b, r, c)) {
      if (rr === lastRow) {
        return true;
      }
      b[r][c] = 'seen';
      stack.push([rr, cc]);
    }
  }
  return false;
};


export class Board {
  constructor(input) {
    this.board = input.map(row => [...row.replace(/\s+/g, '')]);
  }

  winner() {
    if (isWinner('O', this.board)) return 'O';
    if (isWinner('X', transpose(this.board))) return 'X';
    return '';
  }
}
